kwalitee/cli.py

```diff
@@ -31,14 +34,29 @@ def __init__(self, config_file):
                 self.repos.append(Repo(self.config, url, official, compliant_cli))
 pass_ctx = click.make_pass_decorator(CliCtx)
 
+
 @click.group()
 @click.option('-c', '--config-file', help="", default=resource_filename(__name__, 'config.yml'))
 @ocrd_loglevel
 @click.pass_context
-def cli(ctx, config_file, **kwargs): # pylint: disable=unused-argument
+def cli1(ctx, config_file, **kwargs): # pylint: disable=unused-argument
+    """
+    This group creates a list of all Repo instances.
+    """
     ctx.obj = CliCtx(config_file)
 
-@cli.command('clone', help='''
+
+@click.group()
+def cli2():
+    """
+    This group is necessary for all cases where we don't need the list of Repo instances created
```

Comment:
I don't understand why there is a need for a second, context-free click group. I have also never had the use for the
 https://click.palletsprojects.com/en/8.1.x/commands/#merging-multi-commands pattern. If you think the performance/UX penalty of having the `--config-file` option for all commands is too high, I propose to create separate subcommands, one with the context object and one without.

But I really think that `--config-file` should be a global option, the performance hit of parsing a JSON file is negligible.

Reply:
It's not about the `--config-file` option, but about the class `CliCtx`/ the definition of `cli1`:

```python
def cli1(ctx, config_file, **kwargs): # pylint: disable=unused-argument
    """
    This group creates a list of all Repo instances.
    """
    ctx.obj = CliCtx(config_file)
```

If we used this function for every of our options, we create a list of `Repo` instances via `CliCtx` when executing a command – regardless of the function executed needing this list. This is unnecessary. I also don't understand why a command like `validate` should have a `--config-file` option when there is nothing to be configured. Could you elaborate on your reasoning?

Reply:
Normally, I create click applications with nested subcommands, like `ocrd` -> `ocrd workspace` -> `ocrd workspace clone`, where the subcommand takes options common to that subcommand, like `--directory` or `--mets` in the case of `ocrd workspace`. This means that every subcommand is a dedicated `@click.group` which get `add_command`ed to the main CLI.

What you are proposing is to have two `@click.group` that take different options, depending on whether they need the repo list, and merge them on the same level. I think what you want is to have `ocrd-kwalitee` to take the `--config-file` option only if the subcommand is `clone`, `ocrd-tool`, `pull` or `json`. But AFAICT that does not work. I.e.

```
$ ocrd-kwalitee --help
Usage: ocrd-kwalitee [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.
```

shows no `--config-file` option and

```
$ ocrd-kwalitee --config-file foo json
Usage: ocrd-kwalitee [OPTIONS] COMMAND [ARGS]...
Try 'ocrd-kwalitee --help' for help.

Error: No such option: --config-file
```

So instead of merging all the commands on the same level, I think it would be better to have separate `@click.groups`,  one for managing repos (and taking the `--config-file` option), one for `releases` and one for `validate` (the latter two without that option). The help would look like this:

```
$ ocrd-kwalitee --help
Usage: ocrd-kwalitee [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.

Commands:
  repo         Managing repos and related info
  releases   Generate JSON for ocrd_all releases
  validate   Validate created JSON files.

$ ocrd-kwalitee repo --help
Usage: ocrd-kwalitee repo [OPTIONS] COMMAND [ARGS]...

Options:
  --help  Show this message and exit.
  --config-file Location of the config file

Commands:
  clone      Clone all repos
  json       Generate JSON
  ocrd-tool  Return one big list of ocrd tools
  pull       Pull all repos
```

Reply:
Thanks for your feedback. I implemented your suggestion in the latest commit (https://github.com/OCR-D/kwalitee-dashboard-back-end/pull/23/commits/f3fc9a4bcdcee5b58fb1c40497e6284620797c12).

Reply:
LGTM!